manual/tools/caml_tex2.ml

```diff
@@ -34,6 +34,127 @@ let (~!) =
       memo := (key, data) :: !memo;
       data
 
+(** The Output module deals with the analysis and classification
+    of the interpreter output and the parsing of status-related options
+    or annotations for the caml_example environment *)
+module Output = struct
+
+  (** Interpreter output status *)
+  type status =
+    | Ok
+    | Warning of int
+    | Error
+
+  type kind = Annotation | Option
+
+  (** Pretty printer for status *)
+  let pp_status ppf = function
+    | Error -> Printf.fprintf ppf "error"
+    | Ok -> Printf.fprintf ppf "ok"
+    | Warning n -> Printf.fprintf ppf "warning %d" n
+
+  (** Pretty printer for status preceded with an undefined determinant *)
+  let pp_a_status ppf = function
+    | Error -> Printf.fprintf ppf "an error"
+    | Ok -> Printf.fprintf ppf "an ok"
+    | Warning n -> Printf.fprintf ppf "a warning %d" n
+
+  (** {2 Exceptions } *)
+  exception Parsing_error of kind * string
+  type unexpected_report = {source:string; expected:status; got:status}
+  exception Unexpected_status of unexpected_report
+
+  let print_unexpected {source; expected; got} =
+    if expected = Ok then
+      Printf.eprintf
+        "Error when evaluating a caml_example environment in %s:\n\
+         unexpected %a.\n\
+         If %a status was expected, add an [@@expect %a] annotation.\n"
+        source
+        pp_status got
+        pp_a_status got
+        pp_status got
+    else
+      Printf.eprintf
+        "Error when evaluating a guarded caml_example environment in %s:\n \
+         %a status was expected, got %a status.\n\
+         If %a states was in fact expected, change the status annotation to\
+         [@@expect %a].\n"
+        source
+        pp_a_status expected
+        pp_a_status got
+        pp_a_status got
+        pp_status got;
+    flush stderr
+
+  let print_parsing_error k s =
+    match k with
+    | Option ->
+        Printf.eprintf
+          "Unknown caml_example option: [%s].\n\
+           Supported options are \"ok\",\"error\", or \"warning=n\" (with n \
+           a warning number).\n" s
+    | Annotation ->
+        Printf.eprintf
+          "Unknown caml_example phrase annotation: [@@expect %s].\n\
+           Supported annotations are [@@expect ok], [@@expect error],\n\
+           and [@@expect warning n] (with n a warning number).\n" s
+
+
+  (** {2 Output analysis} *)
+  let catch_error s =
+    if string_match ~!{|Error:|} s 0 then Some Error else None
+
+  let catch_warning s =
+    if string_match ~!{|Warning \([0-9]+\):|} s 0 then
+      Some (Warning (int_of_string @@ matched_group 1 s))
+    else
+      None
+
+  let status s = match catch_warning s, catch_error s with
+    | Some w, _ -> w
+    | _, Some e -> e
```

Comment:
Ah, I read another instance of the same pattern below as implementing a priority order. Don't bother, then.